nemo/collections/nlp/models/machine_translation/megatron_nmt_model.py

```diff
@@ -0,0 +1,416 @@
+# Copyright (c) 2021, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import itertools
+import random
+from typing import Optional
+
+import numpy as np
+import torch
+from omegaconf.dictconfig import DictConfig
+from omegaconf.listconfig import ListConfig
+from pytorch_lightning.trainer.trainer import Trainer
+from sacrebleu import corpus_bleu
+
+from nemo.collections.nlp.models.language_modeling.megatron_lm_encoder_decoder_model import (
+    MegatronLMEncoderDecoderModel,
+)
+from nemo.collections.nlp.models.machine_translation.mt_enc_dec_model import MTEncDecModel
+from nemo.collections.nlp.modules.common.megatron.utils import average_losses_across_data_parallel_group
+from nemo.utils import logging
+
+try:
+    from apex.transformer import parallel_state
+
+    HAVE_APEX = True
+except (ImportError, ModuleNotFoundError):
+    HAVE_APEX = False
+
+
+__all__ = ["MegatronNMTModel"]
+
+
+class MegatronNMTModel(MegatronLMEncoderDecoderModel):
+    """
+    Megatron NMT training
+    """
+
+    def __init__(self, cfg: DictConfig, trainer: Trainer):
```

Comment:
Should we use the existing p-tuning here instead of a custom soft-prompt?

Reply:
Not sure I understand. There is no p-tuning or soft-prompt related code here no?

Reply:
I meant for multi-lingual support. But could be added in a separated PR.